sdcflows/transform.py

```diff
@@ -397,33 +386,37 @@ def grid_bspline_weights(target_nii, ctrl_nii):
         step of approximation/extrapolation.
 
     """
-    shape = target_nii.shape[:3]
-    ctrl_sp = ctrl_nii.header.get_zooms()[:3]
-    ras2ijk = np.linalg.inv(ctrl_nii.affine)
-    origin = nb.affines.apply_affine(ras2ijk, [tuple(target_nii.affine[:3, 3])])[0]
+    sample_shape = target_nii.shape[:3]
+    knots_shape = ctrl_nii.shape[:3]
 
+    # Ensure the cross-product of affines is near zero (i.e., both coordinate systems are aligned)
+    if not np.allclose(np.linalg.norm(
+        np.cross(ctrl_nii.affine[:-1, :-1].T, target_nii.affine[:-1, :-1].T),
+        axis=1,
+    ), 0, atol=1e-3):
+        raise RuntimeError("Image's and B-Spline's grids are not aligned.")
+
+    target_to_grid = np.linalg.inv(ctrl_nii.affine) @ target_nii.affine
     wd = []
-    for i, (o, n, sp) in enumerate(
-        zip(origin, shape, target_nii.header.get_zooms()[:3])
-    ):
-        locations = np.arange(0, n, dtype="float16") * sp / ctrl_sp[i] + o
-        knots = np.arange(0, ctrl_nii.shape[i], dtype="float16")
-        distance = np.abs(locations[np.newaxis, ...] - knots[..., np.newaxis])
-
-        within_support = distance < 2.0
-        d_vals, d_idxs = np.unique(distance[within_support], return_inverse=True)
-        # Calculate univariate B-Spline weight for samples within kernel spread
-        weights_vals = _cubic_bspline(d_vals)[d_idxs].astype("float32")
-
-        # Build csr matrix
-        rows, cols = np.where(within_support)
-        weights = csr_matrix(
-            (weights_vals, (rows, cols)),
-            shape=distance.shape,
-        )
-        wd.append(csr_matrix(weights))
+    for axis in range(3):
+        # 3D ijk coordinates of current axis
+        coords = np.zeros((3, sample_shape[axis]), dtype=dtype)
+        coords[axis] = np.arange(sample_shape[axis], dtype=dtype)
+
+        # Calculate the index component of samples w.r.t. B-Spline knots along current axis
+        x = nb.affines.apply_affine(target_to_grid, coords.T)[:, axis]
+
+        # BSpline.design_matrix requires all x be within -4 and 4 padding
+        # This padding results from the B-Spline degree (3) plus one
+        t = np.arange(-4, knots_shape[axis] + 4, dtype=dtype)
+
+        # Calculate K x N collocation matrix (discarding extra padding)
+        colloc_ax = BSpline.design_matrix(x, t, 3)[:, 2:-2]
```

Comment:
Aren't we adding four on each side? Shouldn't this be:

```suggestion
        colloc_ax = BSpline.design_matrix(x, t, 3)[:, 4:-4]
```

Reply:
This is confusing. `BSpline.design_matrix` already removes 2 knots at each end.

If I don't pad the knots `t` with 4 (instead of just 2), then we get out-of-bounds errors.

I have checked with our former implementation and doing this particular padding and de-padding, the weights we obtain with this and the former method are very close (rtol ~ 1e-4).

Reply:
Okay, sounds good.